jupyterhub/auth.py

```diff
@@ -746,19 +749,19 @@ def authenticate(self, handler, data):
         return username
 
 
-    @run_on_executor
-    def pre_spawn_start(self, user, spawner):
+    def pre_exec_fn(self, user, spawner):
         """Open PAM session for user if so configured"""
         if not self.open_sessions:
             return
         try:
             pamela.open_session(user.name, service=self.service, encoding=self.encoding)
         except pamela.PAMError as e:
             self.log.warning("Failed to open PAM session for %s: %s", user.name, e)
-            self.log.warning("Disabling PAM sessions from now on.")
-            self.open_sessions = False
+            # this will be ineffective in the child process
+            # self.log.warning("Disabling PAM sessions from now on.")
+            # self.open_sessions = False
 
-    @run_on_executor
+    #@run_on_executor
```

Comment:
This needs to come back, btw. It's important that all calls to pam happen off the main thread.

Reply:
The reason I removed it was because it caused this check failed with: `jupyterhub/tests/test_internal_ssl_api.py::test_bad_spawn`.

Reply:
Then that would be an issue that needs to be fixed in the test. This method must be allowed to be async. Making it a blocking method isn't an okay fix for the problem, it just ensures the async behavior isn't tested.

But this is hugely useful info! I've been wondering what the source of these errors has been, and I can look into a fix that doesn't assume an async method is synchronous, if you like.

Reply:
The most recent commit restructures this into a function that runs on a intermediate process. This make the run_on_executor statement unnecessary because it is already running of the main thread.